lib/spack/spack/test/compilers.py

```diff
@@ -3,14 +3,29 @@
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
+import pytest
+
 from copy import copy
 from six import iteritems
 
 import spack.spec
 import spack.compilers as compilers
+import spack.compilers.clang
 from spack.compiler import _get_versioned_tuple, Compiler
 
 
+@pytest.fixture()
+def echo_executable_string(monkeypatch):
```

Comment:
name should indicate that this makes *clang* echo an executable string... Why make this a fixture though?  It's only really used in one place so far and it seems very specific.

Reply:
This has been generalized in 7af4f36 to cover all the supported compilers.

Reply:
It took me a while to figure out how this worked. I think the following would be more simple: add a `_test` parameter to `Clang.default_version`, and use `comp` without wrapping in an `Executable` when `_test=True`

Also: now that the functionality has been generalized, the name could still be more specific, like `set_compiler_exe_output`

Reply:
> Also: now that the functionality has been generalized, the name could still be more specific, like set_compiler_exe_output

:+1: 

> I think the following would be more simple: add a _test parameter to Clang.default_version, and use comp without wrapping in an Executable when _test=True

I am not a huge fan of having test logic mixed with business logic, mainly because it's a distraction when you read code. I tried to write a better docstring for the fixture (and I am open to any suggestion to improve it). Another option would be extracting a method that given a string (the output of the compiler command) detects a version:
```python
def detect_version_from_str(output):
    ...
```
then test that part in isolation. Let me know what you think.

Reply:
> Another option would be extracting a method that given a string (the output of the compiler command) detects a version:

I like that

> I am not a huge fan of having test logic mixed with business logic

As a side note: there's a balance to be struck between

* Affecting the real logic with the test logic
* Having the test logic bend over backwards to reach into the real logic

I felt this was trending toward the latter. For example the fixture only works because of how compiler makes a direct call to `Executable`, which is fragile.

Reply:
> As a side note: there's a balance to be struck between [...] which is fragile.

I agree with everything. Probably the only preference I have that might be different from yours is that in this case I would have chosen a fragile fixture over mixing test and business logic. It doesn't matter anymore as I like the latest solution much better than the other two :smile: